content/md/posts/clojure-which-execution-option-to-use.md

```diff
@@ -0,0 +1,217 @@
+
+
+## clojure.main -M
+
+[`clojure.main` namespace](https://clojure.org/reference/repl_and_main) has been the way Clojure code was run (including a REPL) for most of its history.  This is now evolving with the addition of [clojure.exec](#clojure-exec--X).
+
+`clojure.main/main` function looks for a `-main` function in the given namespace (`-m fully-qualified.namespace`) or if no main namespace is found then a REPL session is run.  This is the approach Leiningen has always used and still uses today.
+
+Running a project using Clojure CLI tools without any additional aliases on the command line
+
+```bash
+clojure -M -m practicalli.sudoku-solver
+```
+
+`-M` flag instructs Clojure CLI tools to use `clojure.main` to run the Clojure code.
+
+`-m` flag is an argument to `clojure.main` which specifies the namespace to search for a `-main` function.
+
+The same approach is taken when including an alias, in this example `:alias/name`.
+
+```bash
+clojure -M:alias/name -m practicalli.sudoku-solver
+```
+
+This command will merge values defined in the alias for `:extra-paths` and `extra-deps` into the call.
+
+> When the command line includes the `-m` flag with a namespace it will over-ride the `:main-opts` value if defined in the alias
+
+
+Adding a `:project/run` alias to the project `deps.edn` file provides a simpler way to run the project on the command line
+
+```clojure
+  :project/run
+  {:main-opts ["-m" "practicalli.sudoku-solver"]}
+
+```
+
+Now the project code can be run using the simple command line form
+
+```bash
+clojure -M:project/run
+```
+
+
+### Chaining aliases together
+
+Alises can be used together by chaining their names on the command line
+
+```bash
+clojure -M:env/dev:env/test:lib/cider:repl/rebel
+```
+
+Using several aliases on the command line will merge their `:extra-paths` and `:extra-deps` values.
+
+The `:main-opts` values are not merged and the value from the last `:main-opts` in the alias chain is used with `clojure.main`
+
+Again, if the command line includes the `-m` flag with a namespace, then that namespace is passed to `clojure.main`, ignoring the values from the aliases.
+
+> Aliases can be chained together to merge their configuration, e.g. `clojure -M:env/dev:repl/rebel` will included the configuration from both `:env/dev` and `:repl/rebel`.  With the `-M` flag, only the `:main-opts` from the last alias in the chain is used.  If `-m namespace/name` is also part of the command, this will be used instead of any values from alias `:main-opts` keys.
+
+
+## clojure.exec -X
+
+`-X` flag provides the flexibility to call any fully qualified function, so Clojure code is no longer tied to `-main`
+
+
+
+Arguments are passed as EDN key value pairs and can therefore be included in any order
```

Comment:
This makes it sound like the invoked function can have multiple arguments: it is called with just a single argument, a hash map of all the key/value pairs from `:exec-args` and the command-line merged.